subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/FileParameterUtils.java

```diff
@@ -131,46 +129,39 @@ private static void resolveCompositeOutputFilePropertySpecs(final String ownerDi
                 consumer.accept(new DefaultCacheableOutputFilePropertySpec(propertyName, "." + id, outputFiles, outputType));
             }
         } else {
-            final List<File> roots = Lists.newArrayList();
-            final MutableBoolean nonFileRoot = new MutableBoolean();
             FileCollectionInternal outputFileCollection = fileCollectionFactory.resolving(unpackedValue);
+            AtomicInteger index = new AtomicInteger(0);
             outputFileCollection.visitStructure(new FileCollectionStructureVisitor() {
                 @Override
                 public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
-                    Iterables.addAll(roots, contents);
+                    for (File content : contents) {
+                        FileCollectionInternal outputFiles = fileCollectionFactory.fixed(content);
+                        consumer.accept(new DefaultCacheableOutputFilePropertySpec(propertyName, "$" + (index.incrementAndGet()), outputFiles, outputType));
+                    }
                 }
 
                 @Override
                 public void visitGenericFileTree(FileTreeInternal fileTree) {
-                    nonFileRoot.set(true);
+                    DeprecationLogger.nagUserOfDeprecatedThing("What the hell is this, don't use it as on output. It is currently ignored");
```

Comment:
So this will become an error instead, right?

Reply:
It is an error now.